org.eclipse.xtext.common.types.ui/src/org/eclipse/xtext/common/types/ui/notification/JavaBuilderState.java

```diff
@@ -0,0 +1,259 @@
+		IPath packagePath = resource.getProjectRelativePath();
+		int srcPathSegmentCount = getPackageFragmentRoot(packageFragment).getResource().getProjectRelativePath().segmentCount();
+		for (Object typeLocator : references.keyTable) {
+			if (typeLocator instanceof String) {
+				IPath typeLocatorPath = packageFragment.getJavaProject().getProject().getFile((String) typeLocator)
+						.getProjectRelativePath();
+				if (packagePath.isPrefixOf(typeLocatorPath)) {
+					IPath qualifiedPath = typeLocatorPath.removeFirstSegments(srcPathSegmentCount).removeFileExtension();
+					String typePackageName = qualifiedPath.removeLastSegments(1).toString().replace("/", ".");
+					if (packageName.equals(typePackageName)) {
+						String simpleTypeName = qualifiedPath.lastSegment().toString();
+						qualifiedTypeNames.addAll(
+								getQualifiedTypeNames((String) typeLocator, packageName, simpleTypeName, packageFragment.getJavaProject()));
+					}
+				}
+			}
+		}
+
+		return qualifiedTypeNames;
+	}
+
+	/**
+	 * @return The {@code TypeNames} which are direct children of a given CompilationUnit.
+	 */
+	protected TypeNames _getQualifiedTypeNames(ICompilationUnit compilationUnit) {
+		return getQualifiedTypeNames(getTypeLocator(compilationUnit), getPackageName(compilationUnit),
+				getSimplePrimaryTypeName(compilationUnit), compilationUnit.getJavaProject());
+	}
+
+	private IPackageFragmentRoot getPackageFragmentRoot(IJavaElement javaElement) {
+		IJavaElement parent = javaElement.getParent();
+		return parent instanceof IPackageFragmentRoot ? ((IPackageFragmentRoot) parent) : getPackageFragmentRoot(parent);
+	}
+
+	private TypeNames getQualifiedTypeNames(String typeLocator, String packageName, String simpleName, IJavaProject project) {
+		TypeNames qualifiedTypeNames = new TypeNames(project);
+		String primaryTypeFqn = getQualifedTypeName(packageName, simpleName);
+
+		char[][] typeNames = null;
+		if (state != null) {
+			typeNames = state.getDefinedTypeNamesFor(typeLocator);
+		}
+
+		if (typeNames == null) {
+			TypeNames newTypeNames = new TypeNames(project);
+			newTypeNames.addTypeName(primaryTypeFqn, primaryTypeFqn);
+			return newTypeNames;
+		}
+
+		for (char[] typeName : typeNames) {
+			String qualifiedTypeName = getQualifedTypeName(packageName, new String(typeName));
+			qualifiedTypeNames.addTypeName(qualifiedTypeName, primaryTypeFqn);
+		}
+
+		return qualifiedTypeNames;
+	}
+
+	private String getQualifedTypeName(String packageName, String simpleTypeName) {
+		return packageName == null ? simpleTypeName : packageName + "." + simpleTypeName;
+	}
+
+	private String getTypeLocator(ICompilationUnit compilationUnit) {
+		IResource resource = compilationUnit.getResource();
+		if (resource != null && resource.getProjectRelativePath() != null) {
+			return resource.getProjectRelativePath().toString();
+		}
+		return null;
+	}
+
+	private String getPackageName(ICompilationUnit compilationUnit) {
+		IJavaElement parent = compilationUnit.getParent();
```

Comment:
`org.eclipse.jdt.internal.compiler.env.ICompilationUnit.getPackageName()`

Reply:
We are using the `ICompilationUnit` from the `org.eclipse.jdt.core` package. It doesn't provide such a method.

Reply:
True.